### PR TITLE
Timestamp option & refactor checks

### PR DESCRIPTION
**Pull request by nritsche**
This is a lot, I am sorry...

I think the commit messages explain the changes, but here is an overview:
- Add endpoint config files
- Add a config option that makes an endpoint save a timestamp to the state
- Allow all internal/external forwards, before and after to have the same options.
- Refactor checks on forward replies into a new class

### CODE REVIEW

**Comment by nritsche** [comment]
I still need to resolve the conflicts and make sure the offline test still works...

File: tests/test_on_failure.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
         "call": {
             "forward": {
                 "name": "status",
-                "reply": {"ok": {"type": "bool"}},
+                "reply": {"type": {"ok": "bool"}},
```
Comment (tristpinsm) [inline tests/test_on_failure.py]: For completeness we might want to test the `value` and `identical` checks too.

File: tests/simulate-chime/test_client.py [hunk 2]
```diff
@@ -223,7 +243,14 @@ def test_client():
...
         assert conf["frb"]["update_beam_offset"]["beam_offset"] == 10
 
+        # check frb gain dir in config
+        assert conf["frb_gain"]["frb_gain_dir"] == "insert/sth/useful"
+
+        # check pulsar gain dir in config
+        assert conf["pulsar_gain"]["pulsar_gain_dir"] == ["insert/sth/useful"]
+
     # TODO: check receiver config: bad inputs, gains
+    # TODO check status: timestamp of frb-gain-dir
```
Comment (tristpinsm) [inline tests/simulate-chime/test_client.py]: This is already tested in `test_wait.py`, no?

File: coco/check.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        if not self.save_to_state:
+            return
+        merged = dict()
+        for r in reply.values():
+            if isinstance(r, dict):
+                merged.update(r)
+        self.state.write(self.save_to_state, merged)
+
+
+class ReplyCheck(Check):
```
[the remark this answers is not in the document]
Comment (nritsche) [reply coco/check.py]: Yes, for some reason I though there would be checks on other things than the reply. Maybe the status code.

File: coco/exception.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
+"""Exceptions for coco."""
+
+
+class CocoException(Exception):
+    """Base class for coco exceptions."""
+
+    pass
+
+
+class CocoConfigError(CocoException):
```
[the remark this answers is not in the document]
Comment (nritsche) [reply coco/exception.py]: I fixed that in most places I think

File: conf/endpoints/start.conf [hunk 5]
```diff
@@ -6,16 +6,11 @@ call:
...
-      endpoint: status
-      value:
-          running: True
-      on_failure:
-          slack:
-              message: "Can't connect to nodes: {$failure_hosts}"
+  name: "status"
+  reply:
+    value:
+    running: True
```
[the remark this answers is not in the document]
Comment (nritsche) [reply conf/endpoints/start.conf]: I don't see what you mean. Is this still relevant?

File: tests/test_on_failure.py (hunk 1 above)
Comment (nritsche) [reply tests/test_on_failure.py]: that's done in test_checks.py

File: tests/simulate-chime/test_client.py (hunk 2 above)
Comment (nritsche) [reply tests/simulate-chime/test_client.py]: This is to test if the config actually works with kotekan.

File: conf/endpoints/start.conf (hunk 5 above)
[the remark this answers is not in the document]
Comment (tristpinsm) [reply conf/endpoints/start.conf]: I thought `running` should be in a subblock under `value`

File: tests/test_on_failure.py (hunk 1 above)
Comment (nritsche) [reply tests/test_on_failure.py]: Oh I forgot to add that file.

**Comment by nritsche** [comment]
test_queue doesn't pass at the moment.

File: tests/test_set_state.py [hunk 6]
```diff
@@ -1,45 +1,166 @@
...
     return coco_runner.Runner(CONFIG, ENDPOINTS)
 
 
 def test_get_state(runner):
     """Test get/set_state."""
-    # State should be empty now
-    response = runner.client(GET_ENDPT_NAME)
-    assert "state" in response
-    assert STATE_PATH in response["state"]
-    assert response["state"][STATE_PATH] == {}
```
Comment (tristpinsm) [inline tests/test_set_state.py]: was getting an empty state removed?

File: coco/check.py [hunk 7]
```diff
@@ -296,3 +296,196 @@ def __init__(self, name, expected_types: Dict, on_failure, save_to_state, forwar
...
+            if not result_:
+                if self.state_paths:
+                    for name in self.state_paths.keys():
+                        logger.debug(
+                            f"/{self._name}: Missing value '{name}' in reply from {host}."
+                        )
+                        failed_hosts.add(host)
+                        result.report_failure(self._name, host, "missing", name)
+                if self.state_path:
+                    logger.debug(f"/{self._name}: Empty reply to /{self.name} from {host}.")
```
[the remark this answers is not in the document]
Comment (nritsche) [reply coco/check.py]: In the multiple path case, we know the names of expected reply fields, so we can print them.

[the remark this answers is not in the document]
Comment (tristpinsm) [reply coco/check.py]: yeah, what I was thinking is that reporting an empty reply is a more accurate and concise description of what occurred than indicating that every field is missing. But either way it's not super important

File: tests/test_set_state.py (hunk 6 above)
Comment (nritsche) [reply tests/test_set_state.py]: The state is not empty anymore at this point. The constructors in `check.py` create the parts of the state configured in the configs.

**Pull request merged**